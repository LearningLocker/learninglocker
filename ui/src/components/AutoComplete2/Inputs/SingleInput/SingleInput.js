import React from 'react';
import DebounceInput from 'react-debounce-input';
import { compose, withHandlers, withState } from 'recompose';
import { componentDidUpdate } from 'react-functional-lifecycle';
import { InputWrapper } from 'ui/components/AutoComplete2/Inputs/styled';

const withFocusState = withState('inputFocused', 'setInputFocused', false);
const withFocusStateHandlers = withHandlers({
  setInputFocusedTrue: props => () => {
    props.onFocus();
    props.setInputFocused(true);
  },
  setInputFocusedFalse: props => () => {
    props.setInputFocused(false);
  }
});

const checkInputFocus = componentDidUpdate(({
  hasFocus,
  selectedOption,
  inputFocused,
  setInputFocusedFalse
}) => {
  // It will rendered the other component (renderOption({ option: selectedOption })), so lose focus.
  const shouldRenderSearch = hasFocus || !selectedOption;

  if (!shouldRenderSearch && inputFocused !== false) {
    setInputFocusedFalse();
  }
});

/**
 * An autocomplete component that can take a single selected option and display it,
 * can also be used to search, will use the selected option's display value as the default
 * will display either the selected option or the search input
 */
const SingleInput = ({
  selectedOption,
  placeholder = 'Choose an option',
  searchString = '',
  hasFocus,
  renderOption,
  parseOption = option => option,
  onChangeSearchString = () => { },
  inputFocused,
  setInputFocusedFalse,
  setInputFocusedTrue
}) => {
  const shouldRenderSearch = hasFocus || !selectedOption;

  const value = inputFocused
    ? searchString
    : parseOption(selectedOption);

  return (
    <InputWrapper isFocused={hasFocus}>
      {shouldRenderSearch &&
        <DebounceInput
          debounceTimeout={377}
          value={value}
          placeholder={placeholder}
          onChange={onChangeSearchString}
          onBlur={setInputFocusedFalse}
          onFocus={setInputFocusedTrue} />
      }
      {!shouldRenderSearch &&
        renderOption({ option: selectedOption })
      }
    </InputWrapper>
  );
};

const enhanced = compose(
  withFocusState,
  withFocusStateHandlers,
  checkInputFocus,
);

export default enhanced(SingleInput);
