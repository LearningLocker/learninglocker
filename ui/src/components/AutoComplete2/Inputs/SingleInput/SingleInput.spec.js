import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import SingleInput from './SingleInput';

describe('SingleInput', () => {
  test('should render', () => {
    const mockRenderOption = jest.fn();

    const singleInput = renderer
      .create(<SingleInput selectedOption="test6" renderOption={mockRenderOption} />)
      .toJSON();

    expect(singleInput).toMatchSnapshot();
  });
});
