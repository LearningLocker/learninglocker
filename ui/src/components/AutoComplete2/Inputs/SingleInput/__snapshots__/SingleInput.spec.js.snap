// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleInput should render 1`] = `
.c0 {
  border-radius: 2px;
  border: 1px solid #ccc;
  overflow: hidden;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  min-height: 36px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 input {
  border: none;
  box-shadow: none;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  height: 36px;
  padding: 0 8px;
  outline-color: #F5AB35;
  text-overflow: ellipsis;
}

<div
  className="c0"
/>
`;
